api/pkg/instrumentation/metrics/metrics.go

```diff
@@ -0,0 +1,68 @@
+package metrics
+
+import (
+	"github.com/gojek/mlp/api/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"time"
+)
+
+// MetricName is a type used to define the names of the various metrics collected in the App.
+type MetricName string
+
+var statusLabels = struct {
+	Success string
+	Failure string
+}{
+	Success: "success",
+	Failure: "failure",
+}
+
+// Collector defines the common interface for all metrics collection engines
+type Collector interface {
+	InitMetrics(histogramMap map[MetricName]*prometheus.HistogramVec)
+	MeasureDurationMsSince(key MetricName, histogramMap map[MetricName]*prometheus.HistogramVec, starttime time.Time, labels map[string]string)
+	// MeasureDurationMs is a deferrable version of MeasureDurationMsSince which evaluates labels
+	// at the time of logging
+	MeasureDurationMs(key MetricName, histogramMap map[MetricName]*prometheus.HistogramVec, labels map[string]func() string) func()
+}
+
+// globalMetricsCollector is initialised to a Nop metrics collector. Calling
+// InitMetricsCollector can update this value.
+var globalMetricsCollector = newNopMetricsCollector()
+
+// Glob returns the global metrics collector
+func Glob() Collector {
+	return globalMetricsCollector
+}
+
+// SetGlobMetricsCollector is used to update the global metrics collector instance with the input
+func SetGlobMetricsCollector(c Collector) {
+	globalMetricsCollector = c
+}
+
+// InitMetricsCollector is used to select the appropriate metrics collector and
+// set up the required values for instrumenting.
+func InitMetricsCollector(enabled bool) error {
```

Comment:
I think this `InitMetricsCollector` can be removed. We can have an `InitNopMetricsCollector` (effectively does nothing) and an `InitPrometheusMetricsCollector` (should take in the histogram map and whatever else we add in the future as an argument). Both of them should eventually do `SetGlobMetricsCollector`. But the decision on which one to call can be left to the caller.

In general, the code in `pkg` should not depend on anything outside. The dependency on `api/log` could be an issue if this module is being used elsewhere. But if we break this function down as suggested above, the logging can happen from the caller.